packages/hub/indexing/document-context.js

```diff
@@ -17,7 +17,7 @@ const qs = require('qs');
 
 module.exports = class DocumentContext {
 
-  constructor({ read, search, routers, schema, type, id, sourceId, generation, upstreamDoc, format, includePaths }) {
+  constructor({ read, search, routers, schema, type, id, sourceId, generation, upstreamDoc, included, format, includePaths }) {
```

Comment:
What's the meaning of this `included`? Is it different from having an `included` inside `upstreamDoc`?

Reply:
This argument list has grown unwieldy. Maybe some of these arguments are things that could be computed inside DocumentContext on demand instead of being passed into it? Maybe some of them can become arguments to specific methods (`documentContext.asSearchDoc({ format, included })`) instead?

We are being limited by the current implementation which is very "do all the work up front at once". All the methods are backed by the monolithic `_buildCachedResponse`. Things will get simpler if we can refactor so that things like `searchDoc` and `pristineDoc` and `realms` compute on demand, which shared caches so that subsequent calls to the other methods don't duplicate effort.

Reply:
So the challenge here is that the documents coming out of ephemeral and the git plugins (and I'm assuming the other writers) don't include the `included` part of the document. I wanted to avoid making deeper changes in all of our writers, and instead have the hub:writers.create just pass the additional includes directly into the DocumentContext so that we could derive the schema for the card.  Although now that I'm typing this, I wonder what would happen if we dropped the index and reindexed ephemeral or the git data source--since the included part was never written to the backing store, we don't actually know what the card schema is....

https://github.com/cardstack/cardstack/blob/master/packages/ephemeral/writer.js#L31-L37
https://github.com/cardstack/cardstack/blob/master/packages/git/writer.js#L78-L84

Maybe the correct way to do this is to write all the included resources in the json:api document to the backing store--not just the main `cards` resource